File: frontend/src/services/voice.ts
import type { Conversation, TurnResult, VoiceConfig, VoiceSession } from "../types";
import { get, post } from "./api";

export const voiceApi = {
  config: () => get<VoiceConfig>("/voice/config"),
  start: (body: { mode?: "inbound" | "callback"; case_id?: string; callback_id?: string; language?: string }) =>
    post<VoiceSession>("/voice/sessions", body),
  turn: (id: string, utterance: string) => post<TurnResult>(`/voice/sessions/${id}/turn`, { utterance }),
  pushTranscript: (id: string, messages: { role: string; text: string }[], providerConversationId?: string) =>
    post<Conversation>(`/voice/sessions/${id}/transcript`, { messages, provider_conversation_id: providerConversationId }),
  end: (id: string) => post<Conversation>(`/voice/sessions/${id}/end`),
  conversations: () => get<Conversation[]>("/voice/conversations?limit=30"),
};

File: backend/app/schemas/voice.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class StartSessionIn(BaseModel):
    mode: str = Field(default="inbound", description="inbound (resident calls) or callback (answering a proactive call)")
    case_id: str | None = None
    callback_id: UUID | None = None
    language: str | None = None


class TurnIn(BaseModel):
    utterance: str = Field(min_length=1, max_length=1000)


class TranscriptMessage(BaseModel):
    role: str
    text: str = Field(max_length=4000)


class TranscriptIn(BaseModel):
    messages: list[TranscriptMessage] = Field(default_factory=list)
    provider_conversation_id: str | None = None


class ConversationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    case_id: UUID | None
    callback_id: UUID | None
    channel: str
    status: str
    provider: str
    language: str
    transcript: list[dict[str, Any]]
    summary: str | None
    started_at: datetime
    ended_at: datetime | None
    duration_seconds: float | None
